R/output.R

```diff
@@ -69,6 +84,45 @@ possibly <- function(.f, otherwise, quiet = TRUE) {
   }
 }
 
+#' @export
+#' @rdname safely
+auto_browse <- function(f) {
+  if (is_primitive(f)) {
+    abort("Can not auto_browse() primitive functions")
+  }
+
+  function(...) {
+    withCallingHandlers(
+      f(...),
+      error = function(e) {
+        # 1: h(simpleError(msg, call))
+        # 2: .handleSimpleError(function (e)  <...>
+        # 3: stop(...)
+        frame <- eval_frame(4)
+        browse_in_frame(frame)
+      },
+      warning = function(e) {
+        if (!identical(getOption("warn"), 2)) {
+          frame <- eval_frame(7)
+          browse_in_frame(frame)
+        }
+      }
+    )
+  }
+}
+
+browse_in_frame <- function(frame) {
+  if (.Platform$GUI == "ESS") {
+    # Workaround ESS issue
+    with_env(frame$env, on.exit({
+      browser()
+      NULL
+    }))
+    return_from(frame)
+  } else {
+    eval(quote(browser()), envir = frame$env)
```

Comment:
Ah, that makes some sort of sense, and switching to `expr_eval()` does work correctly. 

Maybe `eval_raw()`? Or `eval_clean()`?  I doubt these new functions are really going sink in with me until I've updated advanced R to use them.